barretenberg/cpp/src/barretenberg/ecc/scalar_multiplication/scalar_multiplication.cpp

```diff
@@ -371,815 +341,551 @@ void add_affine_points(typename Curve::AffineElement* points,
-                product_state.num_points = static_cast<uint32_t>(num_round_points_per_thread + leftovers);
-                product_state.points = points.data();
-                product_state.point_schedule = thread_point_schedule;
-                product_state.num_buckets = static_cast<uint32_t>(num_thread_buckets);
-                AffineElement* output_buckets = reduce_buckets(product_state, true, handle_edge_cases);
-                Element running_sum;
-                running_sum.self_set_infinity();
-
-                // one nice side-effect of the affine trick, is that half of the bucket concatenation
-                // algorithm can use mixed addition formulae, instead of full addition formulae
-                size_t output_it = product_state.num_points - 1;
-                for (size_t k = num_thread_buckets - 1; k > 0; --k) {
-                    if (__builtin_expect(!product_state.bucket_empty_status[k], 1)) {
-                        running_sum += (output_buckets[output_it]);
-                        --output_it;
-                    }
-                    accumulator += running_sum;
-                }
-                running_sum += output_buckets[0];
-                accumulator.self_dbl();
-                accumulator += running_sum;
-
-                // we now need to scale up 'running sum' up to the value of the first bucket.
-                // e.g. if first bucket is 0, no scaling
-                // if first bucket is 1, we need to add (2 * running_sum)
-                if (first_bucket > 0) {
-                    auto multiplier = static_cast<uint32_t>(first_bucket << 1UL);
-                    size_t shift = numeric::get_msb(multiplier);
-                    Element rolling_accumulator = Curve::Group::point_at_infinity;
-                    bool init = false;
-                    while (shift != static_cast<size_t>(-1)) {
-                        if (init) {
-                            rolling_accumulator.self_dbl();
-                            if (((multiplier >> shift) & 1)) {
-                                rolling_accumulator += running_sum;
-                            }
-                        } else {
-                            rolling_accumulator += running_sum;
-                        }
-                        init = true;
-                        shift -= 1;
+    ASSERT(points.size() == scalars.size());
+    const size_t num_msms = points.size();
+
+    std::vector<std::vector<uint32_t>> msm_scalar_indices;
+    std::vector<ThreadWorkUnits> thread_work_units = get_work_units(scalars, msm_scalar_indices);
+    const size_t num_cpus = get_num_cpus();
+    std::vector<std::vector<std::pair<Element, size_t>>> thread_msm_results(num_cpus);
+    BB_ASSERT_EQ(thread_work_units.size(), num_cpus);
+
+    // Once we have our work units, each thread can independently evaluate its assigned msms
+    parallel_for(num_cpus, [&](size_t thread_idx) {
+        if (!thread_work_units[thread_idx].empty()) {
```

Comment:
I would generally think a better algorithm is to not use num_cpus but instead use N where N is the number of chunks that are worth scheduling on a thread, that way we naturally have a work-stealing algorithm and dont have threads ending early / dont need to statically assign them work